scaffolding-chef/lib/scaffolding.sh

```diff
@@ -38,94 +46,193 @@ do_default_unpack() {
+  elif [ -f "$policyfile_rb" ]; then
+    warn "Detected only source $policyfile_rb"
+  else
+    exit_with "Neither $policyfile_lock nor $policyfile_rb was found"
+  fi
 }
 
-do_default_build() {
-  if [ -d "$PLAN_CONTEXT/../policyfiles" ]; then
-    _policyfile_path="$PLAN_CONTEXT/../policyfiles"
+scaffolding_build_policy() {
+  if [ -f "$policyfile_lock" ]; then
+    build_line "Using lock from: $policyfile_lock"
+    chef install "$policyfile_lock"
+  elif [ -f "$policyfile_rb" ]; then
+    warn "Using source from: $policyfile_rb"
+    chef install "$policyfile_rb"
   else
-    if [ -d "$PLAN_CONTEXT/../../policyfiles" ]; then
-      _policyfile_path="$PLAN_CONTEXT/../../policyfiles"
-    else
-      if [ -d "$PLAN_CONTEXT/../../../policyfiles" ]; then
-        _policyfile_path="$PLAN_CONTEXT/../../../policyfiles"
-      else
-        echo "Cannot detect a policyfiles directory!"
-        exit 1
-      fi
-    fi
+    exit_with "Policy: $scaffold_policy_name is not found"
   fi
-  rm -f "$_policyfile_path"/*.lock.json
-  policyfile="$_policyfile_path/$scaffold_policy_name.rb"
-  for x in $(grep include_policy "$policyfile" | awk -F "," '{print $1}' | awk -F '"' '{print $2}' | tr -d " "); do
-    chef install "$_policyfile_path/$x.rb"
-  done
-  chef install "$policyfile"
 }
 
-do_default_install() {
-  chef export "$_policyfile_path/$scaffold_policy_name.lock.json" "$pkg_prefix"
+scaffolding_build_default_toml() {
+  cat << EOF > "$pkg_prefix/default.toml"
+env_path_prefix = "/sbin:/usr/sbin:/usr/local/sbin:/usr/local/bin:/usr/bin:/bin"
+
+interval = 1800
+splay = 180
+log_level = "warn"
+
+named_run_list = ""
+
+ssl_verify_mode = ":verify_none"
+
+[data_collector]
+enable = false
+token = "set_to_your_token"
+server_url = "set_to_your_url"
+
+# Chefs' Node attributes
+[node]
+policy_name = "$scaffold_policy_name"
+policy_group = "$scaffold_policy_group"
+EOF
+
+  chmod 0640 "$pkg_prefix/default.toml"
+}
+
+scaffolding_build_config() {
+  if [ ! -d "$pkg_prefix/config" ]; then
+    mkdir -p "$pkg_prefix/config"
+    chmod 0750 "$pkg_prefix/config"
+  fi
+}
+
+# This config can be used as is without habitat supervisor running
+scaffolding_build_config_bootstrap() {
+  cat << EOF > "$pkg_prefix/config/bootstrap-config.rb"
+policy_name '$scaffold_policy_name'
+policy_group '$scaffold_policy_group'
+
+use_policyfile true
+policy_document_native_api true
 
-  mkdir -p "$pkg_prefix/config"
-  chown 0755 "$pkg_prefix/config"
-  cat << EOF >> "$pkg_prefix/.chef/config.rb"
 cache_path "$pkg_svc_data_path/cache"
 node_path "$pkg_svc_data_path/nodes"
 role_path "$pkg_svc_data_path/roles"
 
 ssl_verify_mode :verify_none
-chef_zero.enabled true
-EOF
 
-  cp "$pkg_prefix/.chef/config.rb" "$pkg_prefix/config/bootstrap-config.rb"
-  cat << EOF >> "$pkg_prefix/config/bootstrap-config.rb"
 ENV['PATH'] = "/sbin:/usr/sbin:/usr/local/sbin:/usr/local/bin:/usr/bin:/bin:#{ENV['PATH']}"
+ENV['GEM_PATH'] = "$pkg_prefix:#{ENV['GEM_PATH']}"
```

Comment:
I don't think we want to treat GEM_PATH the same as PATH.  Chef will require commands from the host system, so we need to let that "leak" into our environment,  but the gems for the policy should remain isolated from any system or external gems that might exist to prevent unintended behavior.

Reply:
My use case here is fix somehow the problem with cookbook gems. 
If I need a gem for a cookbook, it is impossible to install this gem if I offline, or if gem requires some native extensions.
Scaffolding gives you a chance to embed your gems to the policy itself with all required stuff.
Second step is pass new gem path to the embedded chef-client. Chef client itself already has its gems in gem_path.
As result you'll have chefs' embedded gems + policies' embedded gems.

Not sure what I can change here...

Reply:
@smacfarlane The intended result—gems needed by the cookbooks used by this policy are included in the package—makes sense to me. And for them to be found, I think the path into which they were installed needs to be added to the `GEM_PATH`. Maybe this isn't the right implementation, but I see the value in supported a pattern for vendoring into the package whatever gem dependencies the cookbooks have.

Reply:
You're right,  I've talked myself through it a few times now and with @robbkidd and I think that this is probably how it will need to be solved in the current state of Policyfiles. @robbkidd  has some ideas that I'll leave to him to explain, so that I don't mis-communicate what his thoughts are.